### ISSUE TITLE
Feil: Jøkul har flere store utfordringer med Forced Colors-modus (tidligere high contrast mode)

### ISSUE THREAD

**Issue by wkillerud** [fremtind/jokul#2812]
### Bakgrunn

_Forced Colors_-modus er et tilgjengelighetsverktøy som kanskje er mest kjent som _high contrast mode_ fra Windows. Med dette skrudd på overstyrer operativsystemet og nettleseren farger på nettleseren til hva brukeren selv har valgt. Fargetemaene kan være individuelt skreddersydde for den enkeltes syn.

En god artikkel om temaet er Eric Bailey sin [Windows High Contrast Mode, Forced Colors Mode and CSS Custom Properties](https://www.smashingmagazine.com/2022/03/windows-high-contrast-colors-mode-css-custom-properties/).

MDN har en kort oppsummering om [forced-colors](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/forced-colors) med to punktlistes det er viktig å ha i tankene som CSS-utvikler.

Disse fargene overstyres av temaet i `forced-colors`:

- `color`
- `background-color`
- `text-decoration-color`
- `text-emphasis-color`
- `border-color`
- `outline-color`
- `column-rule-color`
- `-webkit-tap-highlight-color`
- SVG sin `fill` og `stroke`

Disse propertiene påvirkes av `forced-colors`

- **`box-shadow` settes til `none`** – denne biter oss spesielt her i Jøkul, hvor vi har brukt box-shadow mye som erstatter for border og outline
- `text-shadow` settes til `none`
- `background-image`
- `color-scheme` settes til `light dark`
- `scrollbar-color` settes til `auto`

I tillegg treffer en browser-bug i Chrome oss der vi bruker `currentColor` for å sette farge, som vi gjør for å få riktig farge i lyst og mørkt tema. Mer om det i Melanie Richards sin post [CurrentColor SVG in forced colors modes](https://melanie-richards.com/blog/currentcolor-svg-hcm/). Essensen er at vi må ha en workaround fram til bugen blir rettet i Chrome.

### Feil i Jøkul

Omtrent hver eneste komponent har én eller annen form for mangel i denne modusen. Testet på Windows 11 med Forced Colors skrudd på. Chrome 99, med kjent bug for SVGer som bruker currentColor.

Temaet jeg testet med:

- Bakgrunn (`Canvas`) satt til `#000000`
- Tekst (`CanvasText`) satt til `#ff18fa`
- Hyperkobling (`LinkText`) satt til `#8080ff`
- Inaktiv tekst (`GrayText`) satt til `#a6a6a6`
- Merket tekst forgrunn (`HighlightText`) satt til `#2b2b2b`
- Merket tekst bakgrunn (`Highlight`) satt til `#d6b4fd`
- Knappetekst (`ButtonText`) satt til `#ffee32`
- Knappebakgrunn (`ButtonFace`) satt til `#1c1c1c`

#### Accordion

- Piler får ikke knappefarge og har derfor dårlig kontrast mot bakgrunnen
- AccordionItem har ingen synlig fokus state med tastatur

#### AlertMessage

- "Typeikon" (til venstre, f. eks. trekanten) får ikke tekstfarge og har derfor dårlig kontrast mot bakgrunnen
- Lukk-knapp får ikke knappefarge og har derfor dårlig kontrast mot bakgrunnen, og gjør seg heller ikke til kjenne som klikkbar
- Lukk-knapp har ingen fokusring vet tastaturnavigasjon
- Gjelder for alle varianter

#### Buttons

- Knapper med loader, loaderen får ikke knappefarger og synes ikke
- Knappene får ikke noen ny border ved fokus, men vokser. Vurder å øke størrelsen på border som en ekstra indikator.
- PrimaryButton som en _lenke_ har ingen visuell forskjell. Det er jo ingen Button, så får ikke ButtonFace bakgrunn slik som en PrimaryButton som _er_ en button.Kanskje knapper-som-lenker bør stiles som knapper i høykontrastmodus?
- Tertiærknapp har en tykk border rundt alle sidene uten padding på høyre og venstre side. Øk paddingen noe til høyre og venstre i forced color modus.

#### Kort

- Ingen av kortene har noen border rundt seg by default. Burde de det, for å være "kort-ish" også i forced color mode?
- Navigasjonskort får lenkefarge, og fokusring ved tastaturnavigasjon som det skal.

#### Checkbox

- Total fail. Selve checkboxen forsvinner, bare labels står igjen.
- Ingen fokusring eller noen som helst indikasjon på tastaturfokus.
- Ingen hover state.
- Fargen på ikonet til feilmeldingsteksten har ikke tekstfarge, og forsvinner med dårlig kontrast mot bakgrunnen.

#### CookieConsent

- Det grønne ikonet for Nettsidene skal fungere teknisk er nettop det - grønt. Skal følge tekstfargen for å være sikker på at kontrasten er god nok.

#### DatePicker

- Ganske ille. Skjemafeltet har ingen border, men en placeholder "avslører" at det er noe der.
- Kalenderpopupen har ingen border som skiller den fra bakgrunnen.
- Kalenderknappen inni skjemafeltet har ikke knappefarge, og heller ikke tastaturfokusring.
- Inni kalenderen når den er åpnet er det ingen fokusring når jeg flytter markeringen med tastatur.
- Fargen på ikonet til feilmeldingsteksten har ikke tekstfarge, og forsvinner med dårlig kontrast mot bakgrunnen.

#### ExpandButton

- Som Accordion, pila får ikke knappefarge og forsvinner med dårlig kontrast mot bakgrunnen
- Ingen synlig forskjell på knappen med og uten tastaturfokus

#### Feedback

- Smileys får ikke knappefarge og forsvinner med dårlig kontrast mot bakgrunnen

#### Footer

- Logo får ikke tekstfarge\* og forsvinner med dårlig kontrast mot bakgrunnen

\*: Jeg vet den merkevaremessig egentlig ikke skal ha andre farger, men her har vi ikke noe valg om vi skal gi brukerne av Forced Color modus en lesbar logo. Uansett om vi overstyrer en bakrunnsfarge rundt SVGen er vi ikke garantert at brukeren klarer å lese det. Det er en grunn til at de bruker denne modusen med akkurat sine valgte farger.

#### Hamburger

- Total fail. Hamburgenen "eksisterer ikke" med mitt tema. Den skal ha knappefarge.
- Hamburger har ingen fokus state ved tastaturnavigering.

#### IconButton

- SVGen får ikke knappefarge og forsvinner med dårlig kontrast mot bakgrunnen.
- Ingen fokusindikator ved tastaturnavigasjon.

#### Icons

- SVGene får ikke tekstfarge og forsvinner med dårlig kontrast mot bakgrunnen

#### Link og NavLink

- De får riktig farge, og en fokusindikator, men fokusindikatoren i lyst tema er noe tynn. Bedre og tydeligere i dark mode. Se om du kan forbedre i lyst tema.

#### Lister

- Diamanten i unummererte lister får ikke tekstfargen fra temaet og kan forsvinne med dårlig kontrast mot bakgrunnsfargen
- Lister med ikoner (check og kryss) får ikke tekstfarge fra temaet. Kan vurdere å bruke GrayText (for disabled tekst) for kryss-ikonet for å skille det mer visuelt fra Check (som bør få tekstfarge, CanvasText).

#### Loader

- Fail. Forsvinner mot bakgrunnen. Må ha tekstfarge (CanvasText).

#### Logo

- Som nevnt i Footer - her må vi ha inn CanvasText. Logoen forsvinner med dårlig kontrast mot bakgrunnsfargen.

#### MessageBox

- Samme problemer som med AlertMessage. Ikonet til venstre forsvinner og skal ha CanvasText-farge. Lukk-knappen må ha ButtonText og fokus-state.

#### RadioButton

- Som Checkbox, total fail. Bare labelen står igjen. Null fokus-state. Null feedback om status. Helt ubrukelig.

#### Select

- Pila i Select-nappen får ikke ButtonText-fargen og er lite synlig, avhengig av tema
  - On second thought: native-varianten stiles _ikke_ som en knapp, men med vanlig CanvasText. Som skjemafelter for øvrig. Kanskje det er den veien vi bør gå – stile alt som CanvasText? En Select vil skille seg ut i et skjema.
  - Autosuggest stiles som et skjemafelt med CanvasText - gå den veien

#### Table

- Klikkbare rader er stilet som CanvasText og gjør seg ikke til kjenne som klikkbare. Stil med ButtonText.
- Markér ved klikk har ingen synlig indikator i Forced Color-modus. Det er for øvrig noe å tenke på ellers også. Få inn et annet tegn på at noe er markert, som f. eks. en sjekkboks foran raden.
- Ekspanderbar tabell, knappens pil har ikke fått ButtonText-styling og forsvinner ved dårlig kontrast mot bakgrunnen. Knappen har ingen fokusindikator.

#### Tabs

- Tabs har en litt rar effekt nå hvor den markerte fanen er den eneste uten noen border under seg. Det _fungerer,_ men er litt merkelig. Gjør så den markerte har en tykkere border.

#### Tag

- Tag baserer seg kun på farge. Innholdet er lesbart, men semantikken forsvinner. Kan vurdere å gjøre noe med borders? Trenger sparre med design her.

#### TextInput

- Ingen border rundt skjemafeltet, det blir usynlig. Bare en label gir et hint om at her er det noe.
- Ingen fokusring.
- Med handling, handlingen får ingen knappestiler og ingen fokusring. Må få ButtonText border og farge på ikon, og en tykkere border eller noe sånt ved fokus.
- Feiltekst har samme problem som alle andre skjemafelt, at ikonet ikke får CanvasText-fargen.

#### TextArea

- Samme problemer som TextInput
- I tillegg, progress-baren synes ikke om man har teller. Selve telleren fungerer, så det er innafor om det viser seg vanskelig å få til. Men prøv en variant som kan fungere for alle.

#### ToggleSwitch

- Bryteren er så og si usynlig, både avskrudd og påskrudd. Må stiles om så ButtonText og ButtonFace får mer spillerom.
- Ingen fokusindikator ved tastaturnavigasjon

#### ToggleSlider

- Knappen er fint synlig, men har ingen fokusindikator ved tastaturnavigasjon

#### Autosuggest

- Mangler border rundt inputfeltet
- Ingen fokus state
- Ikoner får ikke knappestyling eller fokus state

#### useBrowserPreferences

- Lag en lytter og returverdi for forced color mode

#### Feil på dokumentasjonssider

Utenom selve komponentene har vi også noen feil rundtom i portalen.

- Play-knappen til Animasjoner inne under Profil er ikke stilet som en knapp (hardkodet blå, usikker på hvorfor temaet ikke klarer å overstyre)
- Illustrasjonene i spacingskalaen forsvinner, i hvertfall med mitt tema
- Ikonene under profil er hardkodet hvite
- Fokusring ved tastaturnavigasjon mangler helt fra navigasjonsmenyen i toppen.

## Steg for å gjenskape

Dette er best å teste på Windows med Forced Colors skrudd på. Du skrur det på i Innstillinger, under Tilgjengelighet og Kontrasttemaer. Test med Chrome eller Edge for å gjenskape SVG-bugen.

Du kan emulere det i Chrome på macOS siden Chrome 98, men du kan ikke velge egne farger, som gjør det vanskeligere å oppdage problemer som de med SVG. I DevTools, velg `...`-menyen, finn More tools, og velg Rendering. Scroll ned til der du har en samling select-bokser. Sett `forced-color` til `active`. Denne fungerer godt nok til å finne de groveste feilene, men det aller beste er i skrivende stund å teste på Windows.

## Forventet oppførsel

Jeg **må** kunne bruke Jøkul-komponenter i Forced Colors-modus. Et minstekrav er at jeg kan **se** komponentene, og at jeg kan **se hvor tastaturfokus er**. Jeg må kunne bruke **vilkårlige farger i temaet mitt**, så det kan _ikke_ gjøres antagelser om at denne og denne fargen vil kunne fungere på denne og denne bakgrunnen.

## Systeminformasjon

Testet på Windows 11 med Forced Colors skrudd på. Chrome 99, med kjent bug for SVGer som bruker currentColor.

Temaet jeg testet med:

- Bakgrunn (`Canvas`) satt til `#000000`
- Tekst (`CanvasText`) satt til `#ff18fa`
- Hyperkobling (`LinkText`) satt til `#8080ff`
- Inaktiv tekst (`GrayText`) satt til `#a6a6a6`
- Merket tekst forgrunn (`HighlightText`) satt til `#2b2b2b`
- Merket tekst bakgrunn (`Highlight`) satt til `#d6b4fd`
- Knappetekst (`ButtonText`) satt til `#ffee32`
- Knappebakgrunn (`ButtonFace`) satt til `#1c1c1c`

**Comment by piofinn** [comment]
Box-shadow visste vi vel at kom til å bite oss, men hadde ikke forventet at det skulle være _fullt_ så ille 😕 Det meste her kan vel heldigvis fikses ved å erstatte box-shadows med `outline` (for fokusindikatorer) eller `border` (f.eks. for avkrysningsboks/radioknapper), men er fortsatt en kjip jobb. Si fra om du vil fordele litt komponenter, @wkillerud !

Er også verdt å nevne at [`forced-color-adjust`](https://www.w3.org/TR/css-color-adjust-1/#propdef-forced-color-adjust) finnes og er lov å bruke for å gjøre opplevelsen bedre for brukeren i forced-color-modus!

Den SVG-bug'en i Chrome føles som IE-workarounds all over again 😑😑😑 Blæh.

**Comment by wkillerud** [comment]
`forced-colors-adjust` virker som noe som helst bør unngås IMO. Vi skal ha en _veldig_ god grunn til å overstyre en brukers valgte farger. Fun fact: `forced-color-adjust: preserve-parent-color;` er det som må inn i Chrome sin native style for å fikse browserbugen 😅

Takk for tilbudet @piofinn – skal si ifra om jeg går helt lei 😄 Enn så lenge virker det som om mye av det værste kan løses i core med relativt få grep, heldigvis.

**Comment by piofinn** [comment]
Ja, mente ikke å begynne å overstyre farger! Men er vel noen steder der `preserve-parent-color` hadde gjort susen, som du påpeker med Chrome-bugen 😬
